File: src/main/scala/Domain/Trace.scala
package Domain

import com.example.lumina.types.SpanStatus
import com.example.lumina.types.given
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}
import io.circe.{Decoder, Encoder}

import java.time.OffsetDateTime
import java.util.UUID

case class Trace(
    id: UUID,
    agentId: UUID,
    name: String,
    status: SpanStatus,
    startedAt: OffsetDateTime,
    endedAt: Option[OffsetDateTime],
    totalCostUsd: Option[BigDecimal],
    tags: Map[String, String]
)

object Trace {
  given Encoder[Trace] = deriveEncoder[Trace]
  given Decoder[Trace] = deriveDecoder[Trace]
}

File: sdk/typescript/src/API/trace/trace.ts
import { UUID } from "node:crypto"
import { SpanStatus } from "../types"

export interface Trace {
    id: UUID
    agentId: UUID
    sessionId: UUID | null
    name: string
    status: SpanStatus
    startedAt: Date
    endedAt: Date | null
    totalCostUsd: number | null
    tags: Record<string, string>
}

export interface StartTraceBody<T, K> {
    agentId: UUID
    sessionId?: UUID
    name: string
    tags?: Record<string, string>
    input: T
    callback: (input: T) => Promise<K>
}